js/src/components/AdvancedSettings.js

```diff
@@ -0,0 +1,194 @@
+import Collapsible from "./SidebarCollapsible";
+import { __, sprintf } from "@wordpress/i18n";
+import { MultiSelect, Select } from "@yoast/components/src/select/Select";
+import RadioButtonGroup from "@yoast/components/src/radiobutton/RadioButtonGroup";
+import TextInput from "@yoast/components/src/inputs/TextInput";
+import { curryUpdateToHiddenInput, getValueFromHiddenInput } from "@yoast/helpers";
+import { Component } from "@wordpress/element";
+import { Fragment } from "react";
+import { Alert } from "@yoast/components";
+
+/**
+ * Function to check whether the current object refers to a post or a taxonomy.
+ *
+ * @returns {boolean} true if it is a post, false otherwise.
+ */
+const isPost = () => window.wpseoAdminL10n.postType === "post";
+
+/**
+ * The values that are used for the noIndex field differ for posts and taxonomies. This function returns an array of
+ * options that can be used to populate a select field.
+ *
+ * @returns {void} Array Returns an array of options for the noIndex setting.
+ */
+const getNoIndexOptions = () => {
+	const noIndex = window.wpseoAdminL10n.noIndex ? "No" : "Yes";
+	if ( isPost() ) {
+		return [
+			{
+				name: sprintf(
+					__( "Default for %s, currently: %s", "wordpress-seo" ),
+					window.wpseoAdminL10n.label,
+					noIndex,
+				),
+				value: "0",
+			},
+			{ name: __( "No", "wordpress-seo" ), value: "1" },
+			{ name: __( "Yes", "wordpress-seo" ), value: "2" },
+		];
+	}
+	return [
+		{
+			/* Translator: %s translates to the Post Label, %s translates to the indexOption */
+			name: sprintf(
+				__( "Default for %s, currently: %s", "wordpress-seo" ),
+				window.wpseoAdminL10n.label,
+				noIndex,
+			),
+			value: "default",
+		},
+		{ name: __( "Yes", "wordpress-seo" ), value: "index" },
+		{ name: __( "No", "wordpress-seo" ), value: "noindex" },
+	];
+};
+
+/**
+ * Functional component for the Meta Robots No-Index option.
+ *
+ * @returns {Component} The Meta Robots No-Index component.
+ */
+const MetaRobotsNoIndex = () => {
+	const hiddenInputId = isPost ? "#yoast_wpseo_meta-robots-noindex" : "#wpseo_noindex";
+	const value = getValueFromHiddenInput( hiddenInputId );
+	return <Fragment>
+		{
+			window.wpseoAdminL10n.privateBlog &&
+			<Alert type="warning">
+				{ __(
+					"Even though you can set the meta robots setting here, the entire site is set to noindex in the sitewide privacy settings," +
+					"so these settings won't have an effect.",
+					"wordpress-seo"
+				) }
+			</Alert>
+		}
+		<Select
+			label={
+				sprintf(
+					__( "Allow search engines to show this %s in search results?", "wordpress-seo" ),
+					window.wpseoAdminL10n.labelSingular,
+				) }
+			onChange={ curryUpdateToHiddenInput( hiddenInputId ) }
+			name={ "yoast_wpseo_meta-robots-noindex-react" }
+			id={ "yoast_wpseo_meta-robots-noindex-react" }
+			options={ getNoIndexOptions() }
```

Comment:
All data needed to calculate the options is derived from the L10n object, so there is no need to call this function on every render. You can simply define `const metaRobotsNoIndexOptions = getNoIndexOptions()` once before the `MetaRobotsNoIndex`-component is defined. This saves some processing on each render.